[pytorch][ao] Add `torch.matmul` in FloatFunctional/QFunctional

jiaxuzhu92 opened the pull request:
Summary: As title

Test Plan: new unit tests

Differential Revision: D48172841

jerryzh168 commented on torch/ao/nn/quantized/modules/functional_modules.py:
```diff
@@ -79,6 +79,12 @@ def add_relu(self, x: Tensor, y: Tensor) -> Tensor:
         return r
 
+    r"""Operation equivalent to ``torch.matmul(Tensor, Tensor)``"""
+    def matmul(self, x: Tensor, y: Tensor) -> Tensor:
```
can you make corresponding changes in `FXFloatFunctional` as well?

jerryzh168 commented on torch/ao/nn/quantized/modules/functional_modules.py:
```diff
@@ -220,6 +226,12 @@ def add_relu(self, x: Tensor, y: Tensor) -> Tensor:
+    r"""Operation equivalent to ``torch.ops.quantized.matmul(Tensor, Tensor)``"""
+    def matmul(self, x: Tensor, y: Tensor) -> Tensor:
+        r = ops.quantized.matmul(x, y, scale=self.scale, zero_point=self.zero_point)
+        r = self.activation_post_process(r)
```
I feel we probably don't need to call this, could you remove this line and test again to see if it still works?

jiaxuzhu92 replied:
Can you elaborate why we don't need `matmul` in QFunctional? the eager mode convert will swap FloatFunctional to QFunctional right?

jerryzh168 replied:
no I mean we don't need to call `        r = self.activation_post_process(r)`

jerryzh168 commented on torch/ao/nn/quantized/modules/functional_modules.py:
```diff
@@ -220,6 +231,13 @@ def add_relu(self, x: Tensor, y: Tensor) -> Tensor:
+    r"""Operation equivalent to ``torch.ops.quantized.matmul(Tensor, Tensor)``"""
+    def matmul(self, x: Tensor, y: Tensor) -> Tensor:
+        r = ops.quantized.matmul(x, y, scale=self.scale, zero_point=self.zero_point)
+        # Note: this operation is not observed because the observation is not
+        # needed for the quantized op.
```
you can also remove this comment I think, it is observed, but we don't need to call `activation_post_process` in quantized model

jerryzh168 reviewed (approved):
lg, thanks! please see one comment inline

jiaxuzhu92 committed:
[pytorch][ao] Add `torch.matmul` in FloatFunctional/QFunctional (#106831)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/106831

As title

Test Plan: new unit tests

Reviewed By: jerryzh168

Differential Revision: D48172841

fbshipit-source-id: b5e211a21ab8fa32dd7f57a7b820ce96a8d9722d

jiaxuzhu92 commented:
@pytorchbot merge
